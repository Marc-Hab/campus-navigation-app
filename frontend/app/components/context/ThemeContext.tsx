//Context to share state for dark/light theme across all components
import React, { createContext, useState, useEffect, ReactNode } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

// Define the shape of the context data
type ThemeContextType = {
  theme: string;
  toggleTheme: (newTheme: string) => void;
};

// Create the context with a default value of `undefined`
const ThemeContext = createContext<ThemeContextType | undefined>(undefined);

// Define props for the ThemeProvider
type ThemeProviderProps = {
  children: ReactNode;
};

export const ThemeProvider: React.FC<ThemeProviderProps> = ({ children }) => {
  const [theme, setTheme] = useState<string>("light");

  useEffect(() => {
    // Load saved theme from storage
    const getTheme = async () => {
      try {
        const savedTheme = await AsyncStorage.getItem("theme");
        if (savedTheme) {
          setTheme(savedTheme);
        }
      } catch (error) {
        console.log("Error loading theme:", error);
      }
    };
    getTheme();
  }, []);

  const toggleTheme = (newTheme) => {
    setTheme(newTheme);
    AsyncStorage.setItem("theme", newTheme);
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
export default ThemeContext;
