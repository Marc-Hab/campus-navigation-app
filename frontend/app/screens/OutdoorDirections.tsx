import React, { useState, useRef } from "react";
import { View, StyleSheet, TouchableOpacity, Alert, Text } from "react-native";
import { GooglePlacesAutocomplete } from "react-native-google-places-autocomplete";
import OutdoorMap from "../components/maps/OutdoorMap";
import useTheme from "../hooks/useTheme";
import { getStyles } from "../styles";
import { GOOGLE_MAPS_APIKEY } from "../constants";
import CampusPilotHeader from "../components/ui/CampusPilotHeader";
import { Ionicons, MaterialIcons } from "@expo/vector-icons";
import { useLocation } from "../components/context/userLocationContext";
import FontAwesome5 from '@expo/vector-icons/FontAwesome5';

export default function OutdoorDirections() {
  const { theme } = useTheme();
  const globalStyles = getStyles(theme);

  const [startLocation, setStartLocation] = useState(null);
  const [destination, setDestination] = useState(null);
  const [travelMode, setTravelMode] = useState("WALKING");

  // New state variables that update when the user taps GO
  const [submittedStart, setSubmittedStart] = useState(null);
  const [submittedDestination, setSubmittedDestination] = useState(null);

  const { userLocation } = useLocation(); // Get user's current location

  // Refs for GooglePlacesAutocomplete to control input text
  const startLocationRef = useRef();
  const destinationRef = useRef();

  // Update the submitted locations and, in this case, log the navigation info
  const handleGoPress = () => {
    if (startLocation && destination) {
      setSubmittedStart(startLocation);
      setSubmittedDestination(destination);
      console.log(`Navigating from: ${startLocation} to: ${destination} via ${travelMode}`);
    } else {
      Alert.alert(
        "Missing Information",
        "Please fill in both the Start Location and Destination.",
        [{ text: "OK", style: "default" }]
      );
    }
  };

  // Function to set current location in the search bar and state
  const setToCurrentLocation = (type) => {
    const locationText = `${userLocation.latitude.toFixed(6)}, ${userLocation.longitude.toFixed(6)}`;

    if (type === "start") {
      setStartLocation(locationText); // Store as string
      startLocationRef.current.setAddressText(locationText); // Update search bar
    } else if (type === "destination") {
      setDestination(locationText); // Store as string
      destinationRef.current.setAddressText(locationText); // Update search bar
    }
  };

  return (
    <View style={globalStyles.container}>
      {/* Header */}
      <CampusPilotHeader />

      {/* Start Location with Current Location Icon */}
      <View style={styles.inputWithLocationContainer}>
        <GooglePlacesAutocomplete
          ref={startLocationRef}
          styles={{
            container: styles.googleBarWithButton,
            textInput: styles.googleInput,
            listView: { flex: 0 },
          }}
          placeholder="Start Location"
          onPress={(data, details = null) => {
            const latLng = `${details.geometry.location.lat.toFixed(6)}, ${details.geometry.location.lng.toFixed(6)}`;
            setStartLocation(latLng); // Store as string
          }}
          fetchDetails={true}
          query={{
            key: GOOGLE_MAPS_APIKEY,
            language: "en",
          }}
        />

        {/* Current Location Button for Start */}
        <TouchableOpacity onPress={() => setToCurrentLocation("start")} style={styles.locationButton}>
          <MaterialIcons name="my-location" size={36} color="#007BFF" />
        </TouchableOpacity>
      </View>

      {/* Destination with GO Button */}
      <View style={styles.inputWithLocationContainer}>
        <GooglePlacesAutocomplete
          ref={destinationRef}
          styles={{
            container: styles.googleBarWithButton,
            textInput: styles.googleInput,
            listView: { flex: 0 },
          }}
          placeholder="Select Destination"
          onPress={(data, details = null) => {
            const latLng = `${details.geometry.location.lat.toFixed(6)}, ${details.geometry.location.lng.toFixed(6)}`;
            setDestination(latLng); // Store as string
          }}
          fetchDetails={true}
          query={{
            key: GOOGLE_MAPS_APIKEY,
            language: "en",
          }}
        />

        {/* GO Button instead of the current location button */}
        <TouchableOpacity onPress={handleGoPress} style={styles.goButton}>
          <Text style={styles.goButtonText}>GO</Text>
        </TouchableOpacity>
      </View>

      {/* Transportation Mode Selection */}
      <View style={styles.transportModeContainer}>
        {/* Walking Mode */}
        <TouchableOpacity 
          style={[styles.iconButton, travelMode === "WALKING" && styles.iconButtonSelected]}
          onPress={() => setTravelMode("WALKING")}
        >
          <MaterialIcons name="directions-walk" size={24} color={travelMode === "WALKING" ? "#fff" : "#007BFF"} />
        </TouchableOpacity>

        {/* Driving Mode */}
        <TouchableOpacity 
          style={[styles.iconButton, travelMode === "DRIVING" && styles.iconButtonSelected]}
          onPress={() => setTravelMode("DRIVING")}
        >
          <MaterialIcons name="directions-car" size={24} color={travelMode === "DRIVING" ? "#fff" : "#007BFF"} />
        </TouchableOpacity>

        {/* Transit Mode */}
        <TouchableOpacity 
          style={[styles.iconButton, travelMode === "TRANSIT" && styles.iconButtonSelected]}
          onPress={() => setTravelMode("TRANSIT")}
        >
          <MaterialIcons name="directions-transit" size={24} color={travelMode === "TRANSIT" ? "#fff" : "#007BFF"} />
        </TouchableOpacity>
      </View>

      {/* Always Display Map with Submitted Locations */}
      <View style={globalStyles.mapContainer}>
        <OutdoorMap 
          origin={submittedStart} 
          destination={submittedDestination} 
          travelMode={travelMode} 
        />
      </View>
    </View>
  );
}

// Styles
const styles = StyleSheet.create({
  googleBarWithButton: {
    flex: 1,
  },
  googleInput: {
    borderColor: '#ccc',
    borderWidth: 1,
    borderRadius: 10,
    paddingHorizontal: 10,
    height: 50,
    fontSize: 14,
  },
  inputWithLocationContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 10,
    width: '100%',
  },
  locationButton: {
    padding: 10,
  },
  goButton: {
    backgroundColor: '#007BFF',
    padding: 10,
    borderRadius: 10,
    marginLeft: 10,
    justifyContent: 'center',
    alignItems: 'center',
  },
  goButtonText: {
    color: '#fff',
    fontSize: 18,
    fontWeight: 'bold',
  },
  /* Transport Mode Styles */
  transportModeContainer: {
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    marginBottom: 10,
    paddingHorizontal: 20,
  },
  iconButton: {
    alignItems: 'center',
    padding: 3,
    borderRadius: 10,
    borderWidth: 1,
    borderColor: '#007BFF',
    marginHorizontal: 20,
  },
  iconButtonSelected: {
    backgroundColor: '#007BFF',
  },
});
