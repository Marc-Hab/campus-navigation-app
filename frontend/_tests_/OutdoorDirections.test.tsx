import React from "react";
import renderer from "react-test-renderer";
import OutdoorDirections from "@/app/screens/OutdoorDirections";
import { ThemeProvider } from "@/app/components/context/ThemeContext";
import { LocationProvider } from "@/app/components/context/userLocationContext";


jest.mock('@react-native-async-storage/async-storage', () => ({
  setItem: jest.fn(),
  getItem: jest.fn(() => Promise.resolve('light')),
}));

// Mock expo-font properly
jest.mock('expo-font', () => ({
  isLoaded: jest.fn(() => true),
  loadAsync: jest.fn(),
  __metadata__: { loadedNativeFonts: [] }  
}));

// Mock all vector icons
jest.mock('@expo/vector-icons', () => ({
  MaterialIcons: 'MaterialIcons',
  Fontisto: 'Fontisto',
  createIconSet: () => 'Icon'
}));

// Mock GooglePlacesAutocomplete
jest.mock('react-native-google-places-autocomplete', () => ({
  GooglePlacesAutocomplete: () => null
}));

// Mock OutdoorMap component
jest.mock('@/app/components/maps/OutdoorMap', () => {
  return function MockOutdoorMap() {
    return null;
  };
});

test("Outdoor Directions renders correctly", () => {
  const tree = renderer.create(
    <LocationProvider>
      <ThemeProvider>
        <OutdoorDirections />
      </ThemeProvider>
    </LocationProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});