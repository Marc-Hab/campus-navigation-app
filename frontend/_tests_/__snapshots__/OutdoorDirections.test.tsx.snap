// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outdoor Directions renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#000",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "center",
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flex": 0.7,
      }
    }
  >
    <View />
    <View
      style={
        {
          "alignContent": "space-around",
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "padding": 20,
          "width": "100%",
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
          }
        }
      >
        <View>
          <Image
            source={
              {
                "testUri": "../../../assets/images/logo.png",
              }
            }
            style={
              {
                "height": 50,
                "width": 50,
              }
            }
            tintColor="#fff"
          />
        </View>
      </View>
      <View>
        <Text
          style={
            [
              {
                "alignItems": "center",
                "color": "#fff",
                "fontSize": 30,
                "fontWeight": "bold",
                "justifyContent": "space-around",
                "padding": 10,
              },
            ]
          }
        >
          Campus Pilot
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#000",
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
            "width": "100%",
          }
        }
      >
        <View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
          >
            <Image
              source={
                {
                  "testUri": "../../../assets/images/sleep-mode.png",
                }
              }
              style={
                {
                  "height": 40,
                  "width": 40,
                }
              }
              tintColor="#fff"
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      {
        "zIndex": 5,
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <MaterialIcons
          color="#007BFF"
          name="my-location"
          size={36}
        />
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "width": "100%",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#007BFF",
            "borderRadius": 10,
            "justifyContent": "center",
            "marginLeft": 10,
            "opacity": 1,
            "padding": 10,
          }
        }
      >
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 18,
              "fontWeight": "bold",
            }
          }
        >
          GO
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-around",
          "marginBottom": 10,
          "paddingHorizontal": 20,
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#007BFF",
            "borderColor": "#007BFF",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginHorizontal": 20,
            "opacity": 1,
            "padding": 3,
          }
        }
      >
        <MaterialIcons
          color="#fff"
          name="directions-walk"
          size={24}
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#007BFF",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginHorizontal": 20,
            "opacity": 1,
            "padding": 3,
          }
        }
      >
        <MaterialIcons
          color="#007BFF"
          name="directions-car"
          size={24}
        />
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "alignItems": "center",
            "borderColor": "#007BFF",
            "borderRadius": 10,
            "borderWidth": 1,
            "marginHorizontal": 20,
            "opacity": 1,
            "padding": 3,
          }
        }
      >
        <MaterialIcons
          color="#007BFF"
          name="directions-transit"
          size={24}
        />
      </View>
    </View>
  </View>
  <View
    style={
      {
        "borderRadius": 10,
        "flex": 2,
        "width": "100%",
      }
    }
  />
</View>
`;
